src/main/java/net/neoforged/neoforge/transfer/item/ItemResource.java

```diff
@@ -0,0 +1,126 @@
+package net.neoforged.neoforge.transfer.item;
+
+import com.mojang.serialization.Codec;
+import java.util.Optional;
+import net.minecraft.core.Holder;
+import net.minecraft.core.component.DataComponentHolder;
+import net.minecraft.core.component.DataComponentMap;
+import net.minecraft.core.component.DataComponentPatch;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.ItemLike;
+import net.neoforged.neoforge.transfer.Resource;
+
+/**
+ * Immutable combination of an {@link Item} and data components.
+ * Similar to an {@link ItemStack}, but immutable and without amount information.
+ */
+// TODO: probably a "hot take" but I actually prefer ItemVariant :P
```

Comment:
`ItemVariant` does sound more intrinsically correct. `ItemResource` is more generic and doesn't really imply that this thing is actually an identifier-like (or key-like) object.